### PR TITLE
docs: add tutorials for PPIs

### PR DESCRIPTION
**Pull request by gcroci2**
With this PR, I add two tutorial notebooks for PPI data;
-  `data_generation_ppi.ipynb`
   - residue level data generation
   - atomic level data generation
-  `training_ppi.ipynb`
   - classification for GNNs using `GraphDatasets`
   - classification for CNNs using `GridDatasets`
  - references to regression across the notebook

### CODE REVIEW

File: tutorials/training_ppi.ipynb [hunk 1]
```diff
@@ -0,0 +1,6991 @@
...
+    "\n",
+    "        def forward(self, data):\n",
+    "            external_updated1_node_features = self._external1(data.x, data.edge_index, data.edge_attr)\n",
+    "            external_updated2_node_features = self._external2(external_updated1_node_features, data.edge_index, data.edge_attr)\n",
+    "            means_per_graph_external = scatter_mean(external_updated2_node_features, data.batch, dim=0)\n",
+    "            graph_input = means_per_graph_external\n",
+    "            z = self._graph_mlp(graph_input)\n",
+    "            return z\n",
+    "  ```\n",
+    "- `class_weights` is used in classification tasks only, to assign class weights based on the training dataset content. In this case we have a balanced dataset (50% 0 and 50% 1), so it doesn't make sense to use it. It defaults to False. \n",
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:553]:
```suggestion
    "- `class_weights` is used for classification tasks only and assigns class weights based on the training dataset content to account for any potential inbalance between the classes. In this case we have a balanced dataset (50% 0 and 50% 1), so it is not necessary to use it. It defaults to False. \n",
```

File: tutorials/training_ppi.ipynb [hunk 2]
```diff
@@ -0,0 +1,6991 @@
...
+    "- `cuda` and `ngpu` are used for indicating whether to use CUDA and how many GPUs. By default, CUDA is not used and `ngpu` is 0.\n",
+    "- The user can specify a deeprankcore exporter or a custom one in `output_exporters` parameter, together with the path where to save the results. Exporters are used for storing predictions information collected later on during training and testing. We will see later how to read the results saved by `HDF5OutputExporter`."
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:555]: There's a shift in language here from "let's do this and that" to "the user can"

File: tutorials/training_ppi.ipynb [hunk 3]
```diff
@@ -0,0 +1,6991 @@
...
+    "\n",
+    "trainer.configure_optimizers(optimizer, lr, weight_decay)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The default loss function for classification is `torch.nn.CrossEntropyLoss`, while for regression is `torch.nn.MSELoss`. It is also possible to set other PyTorch loss functions by using `Trainer.set_lossfunction` method.\n",
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:616]:
```suggestion
    "The default loss function for classification is `torch.nn.CrossEntropyLoss` and for regression it is `torch.nn.MSELoss`. It is also possible to set some other PyTorch loss functions by using `Trainer.set_lossfunction` method, although not all are currently implemented.\n",
```

File: tutorials/training_ppi.ipynb [hunk 4]
```diff
@@ -0,0 +1,6991 @@
...
+    "\n",
+    "Then we can train our model, using the `train()` method of the `Trainer` class.\n",
+    "\n",
+    "A few notes about `train()` method parameters:\n",
+    "- `earlystop_patience`, `earlystop_maxgap` and `min_epoch` are used for controlling early stopping logic. `earlystop_patience` indicates the number of epochs after which the training ends if the validation loss as not improved. `earlystop_maxgap` indicated the maximum difference allowed between validation and training loss, and `min_epoch` is the minimum epoch to be reached before looking at maxgap.\n",
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:621]:
```suggestion
    "- `earlystop_patience`, `earlystop_maxgap` and `min_epoch` are used for controlling early stopping logic. `earlystop_patience` indicates the number of epochs after which the training ends if the validation loss does not improve. `earlystop_maxgap` indicated the maximum difference allowed between validation and training loss, and `min_epoch` is the minimum number of epochs to be reached before evaluating `maxgap`.\n",
```

File: tutorials/training_ppi.ipynb [hunk 5]
```diff
@@ -0,0 +1,6991 @@
...
+    "- `validate` is set True to perform validation on an independent dataset, which we called `dataset_val` few cells above.\n",
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:622]:
```suggestion
    "- If `validate` is set to `True`, validation is performed on an independent dataset, which we called `dataset_val` few cells above. If set to `False`, validation is performed on the training dataset itself (not recommended). \n",
```

File: tutorials/training_ppi.ipynb [hunk 6]
```diff
@@ -0,0 +1,6991 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### Results visualization\n",
+    "\n",
+    "Finally, we can inspect the results saved by `HDF5OutputExporter`, which can be found in `data/ppi/gnn_classif/` folder in the form of an HDF5 file, `output_exporter.hdf5`. Note that the folder contains the saved pre-trained model as well. \n",
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:758]:
```suggestion
    "Finally, we can inspect the results saved by `HDF5OutputExporter`, which can be found in the `data/ppi/gnn_classif/` folder in the form of an HDF5 file, `output_exporter.hdf5`. Note that the folder contains the saved pre-trained model as well. \n",
```

File: tutorials/training_ppi.ipynb [hunk 7]
```diff
@@ -0,0 +1,6991 @@
...
+    "    width=700, height=400,\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "And now let's print a few metrics of interest for classification tasks: the Area under the ROC curve (AUC), and for a threshold of 0.5 the precision, the recall, the accuracy and the f1 score."
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:3818]:
```suggestion
    "And now let's print a few metrics of interest for classification tasks: the [area under the ROC curve](https://en.wikipedia.org/wiki/Receiver_operating_characteristic#Area_under_the_curve) (AUC), and for a threshold of 0.5 the [precision, recall, accuracy and f1 score](https://en.wikipedia.org/wiki/Precision_and_recall#Definition)."
```

File: tutorials/training_ppi.ipynb [hunk 8]
```diff
@@ -0,0 +1,6991 @@
...
+    "    print(f'\\nMetrics for {set}:')\n",
+    "    fpr_roc, tpr_roc, thr_roc = roc_curve(y_true, y_score)\n",
+    "    auc_score = auc(fpr_roc, tpr_roc)\n",
+    "    print(f'AUC: {round(auc_score, 1)}')\n",
+    "    print(f'Considering a threshold of {threshold}')\n",
+    "    y_pred = (y_score > threshold)*1\n",
+    "    print(f'- Precision: {round(precision_score(y_true, y_pred), 1)}')\n",
+    "    print(f'- Recall: {round(recall_score(y_true, y_pred), 1)}')\n",
+    "    print(f'- Accuracy: {round(accuracy_score(y_true, y_pred), 1)}')\n",
+    "    print(f'- F1: {round(f1_score(y_true, y_pred), 1)}')"
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:3876]: string format the digits rather than rounding (so it looks nice if the trailing digit is a 0) and use 2 digits rather than 1:

```suggestion
    "    print(f'AUC: {auc_score, 1:.2f}')\n",
    "    print(f'Considering a threshold of {threshold}')\n",
    "    y_pred = (y_score > threshold)*1\n",
    "    print(f'- Precision: {precision_score(y_true, y_pred):.2f}')\n",
    "    print(f'- Recall: {recall_score(y_true, y_pred):.2f}')\n",
    "    print(f'- Accuracy: {accuracy_score(y_true, y_pred):.2f}')\n",
    "    print(f'- F1: {f1_score(y_true, y_pred):.2f}')"
```

File: tutorials/training_ppi.ipynb [hunk 9]
```diff
@@ -0,0 +1,6991 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Of course such metrics results are not satisfying from the performance point of view, but we did use only 100 data points, which of course is not enough to train a neural network. Now let's do the same exercise but using grids instead of graphs and CNNs instead of GNNs. "
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:3884]:
```suggestion
    "Note that the poor performance of this network is due to the small number of datapoints used in this tutorial. For a strong network performance we suggest using at least 100 000 data points."
```

File: tutorials/training_ppi.ipynb [hunk 10]
```diff
@@ -0,0 +1,6991 @@
...
+    "\n",
+    "As for graphs, the class `Trainer` is used for training, validation and testing of the PyTorch-based CNN. "
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "- Also in this case, `neuralnet` can be whatever neural network class which inherits from `torch.nn.Module`, and it shouldn't be specific to regression or classification in terms of output shape. Here we use `CnnClassification` (in `deeprankcore.neuralnets.cnn.model3d`), whose architecture is shown below:\n",
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:4018]: see my comment on grid `Trainer` class. 

Maybe add a comment about the optimizer, similar to grid case?

File: tutorials/training_ppi.ipynb [hunk 11]
```diff
@@ -0,0 +1,6991 @@
...
+    "    print(f'\\nMetrics for {set}:')\n",
+    "    fpr_roc, tpr_roc, thr_roc = roc_curve(y_true, y_score)\n",
+    "    auc_score = auc(fpr_roc, tpr_roc)\n",
+    "    print(f'AUC: {round(auc_score, 1)}')\n",
+    "    print(f'Considering a threshold of {threshold}')\n",
+    "    y_pred = (y_score > threshold)*1\n",
+    "    print(f'- Precision: {round(precision_score(y_true, y_pred), 1)}')\n",
+    "    print(f'- Recall: {round(recall_score(y_true, y_pred), 1)}')\n",
+    "    print(f'- Accuracy: {round(accuracy_score(y_true, y_pred), 1)}')\n",
+    "    print(f'- F1: {round(f1_score(y_true, y_pred), 1)}')"
```
Comment (DaniBodor) [inline tutorials/training_ppi.ipynb:6957]: see comment above for grids

**Review by DaniBodor — COMMENTED** [review]
Really great tutorials!
I have mainly some textual suggestions and in a few places make some of the settings more explicit/visible so that users see how/where to find them.

**Review by DaniBodor — APPROVED** [review]

File: tutorials/data_generation_ppi.ipynb [hunk 12]
```diff
@@ -0,0 +1,1285 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The example data used in this tutorial are available on Zenodo at [this record address](https://zenodo.org/record/7997586). To download the data, please visit the link and click the “Download\" button. Unzip the downloaded file, and save the contents as a folder named `data/` in the same directory as this notebook. The name data and the folder location are optional but recommended, as this are the name and the location we will use to refer to the folder throughout the tutorial.\n",
+    "\n",
+    "The dataset contains only 100 data points, which are obviously not enough to develop an impactful predictive model, and the scope of its use is indeed only demonstrative and informative for the users."
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply tutorials/data_generation_ppi.ipynb:34]: I'll upload two distinct folders on Zenodo, one for tutorial 1 and another for tutorial 2. So they can pick whatever they need and it's not mandatory to download the big one. Also I'll remove the MacOS stuff from the new folders.

File: tutorials/data_generation_ppi.ipynb [hunk 13]
```diff
@@ -0,0 +1,1285 @@
...
+      "100 queries added to the collection.\n",
+      "Queries ready to be processed.\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "queries = QueryCollection()\n",
+    "\n",
+    "interface_distance_cutoff = 15  # max distance in Å between two interacting residues/atoms of two proteins\n",
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply tutorials/data_generation_ppi.ipynb:265]: I'll use something different from the default, like 8

**Comment by gcroci2** [comment]
> Really great tutorials! I think the main comment I have is that I would split up the zotero into two separate repo's. One for pdb data and another for hdf5 data. People going through data generation don't need to download 2GB, especially because almost all of it will be deleted in the 1st step anyway.
> 
> The HDF5 repo is the only for people that don't want to go through the process of data generation (actually, that one can just be the current one, with all data because the pdbs don't add much).
> 
> Apart from that, these are mostly some minor textual suggestions and in a few places make some of the settings more explicit/visible so that users see how/where to find them.

- I edited the [data on Zenodo](https://zenodo.org/record/8034819) for them to be in two different folders, one for each tutorial (data_raw and data_processed, respectively).
- I edited the setup - software in both tutorials. I tested a new environment using the `tutorials` extras_require version of the package, and everything works on my machine. 
- I removed the `hdf5explorer` dependency which was causing many issues on my Mac installation (with `pyqt5` in particular, even if the last time I made it work with the links indicated in the readme; apparently it doesn't work anymore). We're not really using it and at the bottom of the readme there this package is mentioned and there is a link to its documentation, which should include the installation as well. Also, I remember that we were thinking to remove `hdf5explorer`, but we decided to leave it at the time. Anyway nobody is using it. 
- I implemented the vast majority of your suggestions, I feel it is much more precise now! Thanks for the precious feedback <3

**Comment by gcroci2** [comment]
If you'll review this PR again when I am on holiday, feel free to merge it if there is nothing else to add @DaniBodor

**Review by DaniBodor — APPROVED** [review]
Nice! 👍 

I fixed a few remaining typos/language issues in the notebooks, apart from that it looks good!

I agree with your point about publishing it with a finished run, so people could choose to gloss over it instead of running the notebook. If we do it that way, I think we should state that they can either gloss or clear the notebook and run it themselves. Shall I add a comment about that and run the notebooks, or do you prefer to leave them blank? Or do you think this will prompt too many users to just read and not try it out themselves?

Also, I noticed that you left code snippets of the neural networks inside the md cells. I still think it's quite distracting from the flow to have such a large code block in the middle of a cell for something that is non-essential to understand at a basic level how to run the software. Did you have a particular reason for leaving it or did you miss my previous comment on this? 

Regarding Zenodo. Maybe a nice idea to already add a .gitignore file with `*` in each folder, just to make extra sure that no data inside is tracked. (this is especially useful in case people save the data in a different location than the recommended one). Also, remove the Mac leftovers (DS_Store thing).

Finally, should there be some explanation of how to use notebooks at all (or at least a link explaining it)? Especially for users who don't use notebooks integrated in their IDE, it would be nice to be able to run it from jupyter-lab (like we do for the workshops). Maybe there could be a README for the notebooks specifically, with the generic info like intro and install instructions that is currently on top of both notebooks (then it doesn't need to be repeated), as well as how to use it from IDE or jupyter-lab. I did a draft of this in PR #455 , see if you like it or not.

I made some changes in (mainly to README) in #452 that are pointing here, because I branched off from here to avoid conflicts in the future. That one can either be merged to this first, or can be merged to main after this goes through.

File: setup.cfg [hunk 14]
```diff
@@ -76,6 +75,11 @@ publishing =
     build
     twine
     wheel
+tutorials = 
+    notebook
+    pytest
+    scikit-learn
```
Comment (DaniBodor) [inline setup.cfg:81]:
```suggestion
```

It's part of the main package already

File: setup.cfg [hunk 15]
```diff
@@ -40,7 +40,6 @@ install_requires =
     numpy >= 1.21.5
     scipy >= 1.8.0
     h5py >= 3.6.0
-    h5xplorer
```
Comment (DaniBodor) [inline setup.cfg:43]: can we also remove it from action.yml?

**Comment by gcroci2** [comment]
Give me a sign when you merge the other two PRs, if there is nothing else to add I'll merge this then :) @DaniBodor

**Comment by DaniBodor** [comment]
It might make sense to wait for #446 to get closed before finalizing this, so we can include those changes (otherwise it can be updated in that PR).

**Comment by gcroci2** [comment]
> It might make sense to wait for #446 to get closed before finalizing this, so we can include those changes (otherwise it can be updated in that PR).

I would update the notebooks in PR #446 directly, together with the rest of the docs
